add new page for built-in functions and standard library (#54)
RSE-Sheffield/pando-python: episodes/optimisation-using-python.md

---
title: "Using Python Language Features and the Standard Library"
teaching: 10
exercises: 5
---

:::::::::::::::::::::::::::::::::::::: questions

- Why are Python loops slower than specialised functions?
- How can I make my code more readable and faster?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Able to utilise Python language features effectively
- Able to search Python documentation for functionality available in built-in types and in the standard library
- Able to identify when Python code can be rewritten to perform execution in the back-end.

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: instructor

This episode discusses relatively fundamental features of Python.

For students experienced with writing Python, many of these points may be unnecessary. However, self-taught students—especially if they have previously studied lower-level languages with a less powerful standard library—may have adopted “unpythonic” habits and will particularly benefit from this section.

::::::::::::::::::::::::::::::::::::::::::::::::

> Before we look at data structures, algorithms and third-party libraries, we should take a few minutes to make sure we’re familiar with the fundamentals of Python.
> If you’ve learned to program in another language, chances are you’ve picked up some habits from that language that don’t work well in Python.
> 
> Most of the bad habits that took me a while to unlearn—and that I’ve observed in others, too—come from learning to program in a lower-level language (like C or Delphi), with a less powerful standard library.

Reviewer: Probably best to avoid first-person "took me a while" (and below "when I was in undergrad")

```suggestion
Before we look at data structures, algorithms and third-party libraries, it's worth reviewing the fundamentals of Python.
If you're familiar with other programming languages, like C or Delphi, you might not know the Pythonic approaches. Whilst you can write Python in a way similar to other languages, it is often more effective to take advantage of Python's principles and idioms.
```

## Built-in Functions

> For example, back when I was in undergrad, if you’d asked me to sum up a bunch of data points, I would have written something like the first function in this code sample:

Reviewer: First person, and specifying the identifier of the first function

```suggestion
For example, you might think to sum a list of numbers by using a for loop, as would be typical in C, as shown in the function `manualSumC()` and `manualSumPy()` below. 
```
This was my first re-draft after removing the comments, but the additional second sentence feels like a redundant spoiler, it is covered below.

```suggestion
For example, you might think to sum a list of numbers by using a for loop, as would be typical in C, as shown in the function `manualSumC()` and `manualSumPy()` below. Whereas using the built-in `sum()` is both clearer and faster!
```

```Python
import random
from timeit import timeit

> N = 100_000  # Number of elements in the list

Reviewer: Consistency with other the code examples, and potentially easier for novices to recognise.

```suggestion
N = 100000  # Number of elements in the list
```

# Ensure every list is the same
random.seed(12)
my_data = [random.random() for i in range(N)]

> def manualSumC():  # my first attempt

Reviewer: First person, also kind of split over this. I don't think other code examples have comments on the actual methods instead they're typically described in the text before/after

```suggestion
def manualSumC():
```

    n = 0
    for i in range(len(my_data)):
        n += my_data[i]
    return n

> def manualSumPy():  # slightly improved

Reviewer: ```suggestion
def manualSumPy(): 
```

    n = 0
    for evt_count in my_data:
        n += evt_count
    return n

> def builtinSum():  # fastest and most readable

Reviewer: ```suggestion
def builtinSum(): 
```

    return sum(my_data)

repeats = 1000
print(f"manualSumC: {timeit(manualSumC, globals=globals(), number=repeats):.3f}ms")
print(f"manualSumPy: {timeit(manualSumPy, globals=globals(), number=repeats):.3f}ms")
print(f"builtinSum: {timeit(builtinSum, globals=globals(), number=repeats):.3f}ms")
```

> Even just replacing the iteration over indices (which may be a habit you’ve picked up if you first learned to program in C) with a more pythonic iteration over the elements themselves speeds up the code by about 2×.

Reviewer: Capitalisation/consistency (I'm not personally sure what that small x is)

```suggestion
Even just replacing the iteration over indices (which may be a habit you’ve picked up if you first learned to program in C) with a more Pythonic iteration over the elements themselves speeds up the code by about 2x.
```

Author: For the ×, I’ll write a separate comment.

Re “Pythonic”, I’m not quite sure. My intuition prefers the lower case variant (since it is an adjective, even though it may be derived from a proper noun), but I’d be okay with upper case as well. However, how would you write “unpythonic” then (which is used elsewhere on this page)? “un-Pythonic” feels awkward to me.

> But even better, by switching to the builtin `sum` function our code becomes about 8× faster, doing the exact same operation!

Reviewer: Consistency, + missing clearer note as I removed the comment from code.

```suggestion
But even better, by switching to the built-in `sum` function our code becomes about 8x faster, and easier to parse, doing the exact same operation!
```

M = 2  # N*M == Range over which the elements span

def generateInputs():
    random.seed(12)  # Ensure every list is the same
    return [random.randint(0, int(N*M)) for i in range(N)]
    
def manualSearch():
    ls = generateInputs()
    ct = 0
    for i in range(0, int(N*M), M):
        for j in range(0, len(ls)):
            if ls[j] == i:
                ct += 1
                break

def operatorSearch():
    ls = generateInputs()
    ct = 0
    for i in range(0, int(N*M), M):
        if i in ls:
            ct += 1

repeats = 1000
gen_time = timeit(generateInputs, number=repeats)
print(f"manualSearch: {timeit(manualSearch, number=repeats)-gen_time:.2f}ms")
print(f"operatorSearch: {timeit(operatorSearch, number=repeats)-gen_time:.2f}ms")
```

This results in the manual Python implementation being 5x slower, doing the exact same operation!

```output
manualSearch: 152.15ms
operatorSearch: 28.43ms
```

An easy approach to follow is that if two blocks of code do the same operation, the one that contains less Python is probably faster. This won't apply if you're using 3rd party packages written purely in Python though.

::::::::::::::::::::::::::::::::::::: callout

> ### Python bytecode

Reviewer: todo @Robadob (me), move the detail of this to a backpage somewhere.

#61

        >>   30 FOR_ITER                10 (to 52)
             32 STORE_FAST               2 (i)

  7          34 LOAD_FAST                2 (i)
             36 LOAD_FAST                0 (ls)
             38 CONTAINS_OP              0
             40 POP_JUMP_IF_FALSE       25 (to 50)

  8          42 LOAD_FAST                1 (ct)
             44 LOAD_CONST               2 (1)
             46 INPLACE_ADD
             48 STORE_FAST               1 (ct)
        >>   50 JUMP_ABSOLUTE           15 (to 30)

  6     >>   52 LOAD_CONST               0 (None)
             54 RETURN_VALUE
```

:::::::::::::::::::::::::::::::::::::::::::::

## Example: Parsing data from a text file

In C, since there is no high-level `string` datatype, parsing strings can be fairly arduous work where you repeatedly look for the index of a separator character in the string and use that index to split the string up.

::::::::::::::::::::::::::::::::::::: challenge

Let’s say we have read in some data from a text file, each line containing a time bin and a mean energy:

```python
f = [
    ' 0000   0.9819 ',
    ' 0001   0.3435 ',
    # ...
    ' 0099   0.2275 ',
    ' 0100   0.7067 ',
    # ...
]
```

> A colleague (who learned to program in C before he started using Python) wrote the following code to parse the data into a dictionary:

Reviewer: 1st/3rd person

```suggestion
If you've a C programming background, you may write the following code to parse the data into a dictionary:
```

```python
def manualSplit():
    data = {}
    for line in f:
        first_char = line.find("0")
        end_time = line.find(" ", first_char, -1)

        energy_found = line.find(".", end_time, -1)
        begin_energy = line.rfind(" ", end_time, energy_found)

>         end_energy = line.find(" ", energy_found, -1)

Reviewer: ```suggestion
        end_energy = line.find(" ", begin_energy, -1)
```

        if end_energy == -1:
            end_energy = len(line)
        
        time = line[first_char:end_time]
        energy = line[begin_energy + 1:end_energy]

        data[time] = energy
    return data
```

Can you find a shorter, more easily understandable way to write this in Python?

> <!-- Did you spot the bug in the manual implementation? ;) -->

Reviewer: Probably best to leave out hidden bugs (other people may be teaching this!).

I'm assuming my suggestion above is the bug, if not please fix.

```suggestion

```

Author: A bit of a story on that:

Originally, when I wrote this comment I was planning to add an off-by-one error in one of the slices, to make the point that this code is not just really hard to read, but also makes it easy to make mistakes. Then I decided against it and forgot to remove this comment.
Then, when I ran the code just now to double check that I _really_ didn’t put in that off-by-one error … I discovered a different bug. 😅🫣

I’ve now pushed a commit that fixed the other bug, removed the misleading comment and instead added a note after the solution to point out another potential weakness of that code.

:::::::::::::::::::::::: hint

> Python strings have a lot of methods to perform common operations, like removing suffixes, replacing substrings, joining or splitting, stripping whitespaces, and much more. See the documentation at https://docs.python.org/3/library/stdtypes.html#string-methods for a full list.

Reviewer: Consistency with other links

```suggestion
Python strings have a lot of methods to perform common operations, like removing suffixes, replacing substrings, joining or splitting, stripping whitespaces, and much more. See Python's [string methods documentation](https://docs.python.org/3/library/stdtypes.html#string-methods) for a full list.
```

:::::::::::::::::::::::::::::::::

:::::::::::::::::::::::: solution

```python

def builtinSplit():
    data = {}
    for line in f:
        time, energy = line.split()
        data[time] = energy
    return data
```

This is much more readable.

> The code that’s executed by CPython may use similar indexing steps as in `manualSplit`; however, since this is all happening “under the hood” in C code, it is once again faster. 

Reviewer: Quote consistency with rest of site, likewise i always add empty brackets when I inline function names to clearly denote it's a function.

```suggestion
The code that’s executed by CPython may use a similar approach as in `manualSplit()`; however, since this is all happening "under the hood" in C code, it is once again faster. 
```
